README.md
[![Build Status](https://travis-ci.org/karupanerura/Geo-Hex-V3-XS.svg?branch=master)](https://travis-ci.org/karupanerura/Geo-Hex-V3-XS)
# NAME

Geo::Hex::V3::XS - GeoHex implementation with XS. (c-geohex3 Perl5 binding.)

# SYNOPSIS

    use Geo::Hex::V3::XS;

    my $zone = Geo::Hex::V3::XS->new(code => 'XM488276746');

    # or

    my $zone = Geo::Hex::V3::XS->new(lat => 35.579826, lng => 139.654524, level => 9);

    say 'geohex.code:  ', $zone->code;
    say 'geohex.lat:   ', $zone->lat;
    say 'geohex.lng:   ', $zone->lng;
    say 'geohex.level: ', $zone->level;

# DESCRIPTION

Geo::Hex::V3::XS is [GeoHex](http://geohex.net/) implementation.

# FUNCTIONS

- `my $geohex_code = encode_geohex($lat, $lng, $level)`

    Convert location,level to geohex's code.

- `my ($lat, $lng, $code) = decode_geohex($geohex_code)`

    Convert geohex's code to location,level.
    This location is center of geohex.

- `my $size = geohex_hexsize($level)`

    Calculate geohex hex size by level.

# METHODS

- `my $zone = Geo::Hex::V3::XS->new(...)`

    Create geohex zone object.

    Arguments can be:

    - `code`

        Create geohex zone object from geohex code.

            use Geo::Hex::V3::XS;

            my $zone = Geo::Hex::V3::XS->new(code => 'XM488548');

    - `lat/lng/level`

        Create geohex zone object from location with level.

            use Geo::Hex::V3::XS;

            my $zone = Geo::Hex::V3::XS->new(
                lat   => 40.5814792855475,
                lng   => 134.296601127877,
                level => 7,
            );

- `$zone->lat`

    Get geohex center location latitude.

- `$zone->lng`

    Get geohex center location longitude.

- `$zone->code`

    Get geohex code.

- `$zone->level`

    Get geohex level. (0-15)

- `$zone->size`

    Get geohex size.

- `my @locations = $zone->polygon()`

    Get vertex locations of a geohex polygon.

# LICENSE

Copyright (C) 2015 karupanerura <karupa@cpan.org>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

# ALGORITHM LICENSE

Copyright (c) 2009 @sa2da (http://twitter.com/sa2da)
http://www.geohex.org
Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

# AUTHOR

karupanerura <karupa@cpan.org>


Changes
Revision history for Perl extension Geo-Hex-V3-XS

{{$NEXT}}

0.11 2015-10-01T02:46:46Z

   - Fixed overflow on 32bit environment.

0.10 2015-09-03T06:43:16Z

   - Fixed x/y/level constructor.

0.09 2015-09-03T00:48:30Z

   - Added x/y/level constructor

0.08 2015-08-15T09:25:49Z

   - Upgrade c-geohex3 to 1.40. (improvement of accuracy)

0.07 2015-06-22T23:45:38Z

   - Fixed compile option.

0.06 2015-06-06T01:08:43Z

   - Fixed #1 test expected value for `-Duselongdouble` perl.

0.05 2015-06-03T03:43:00Z

   - Fixed dependency.

0.04 2015-06-03T03:16:10Z

   - No run geohex-3 testing on build phase.

0.03 2015-06-03T03:00:12Z

   - Provides `geohex_hexsize` function.

0.02 2015-05-31T02:07:10Z

    - Fixed license to MIT.

0.01 2015-05-31T01:54:34Z

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use builder::MyBuilder;
use File::Basename;
use File::Spec;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Geo-Hex-V3-XS',
    module_name     => 'Geo::Hex::V3::XS',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw(src)],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,


    needs_compiler_c99 => 1,
    needs_compiler_cpp => 0,
    generate_ppport_h => 'src/ppport.h',
    generate_xshelper_h => 'src/xshelper.h',
    cc_warnings => 1,
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = builder::MyBuilder->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

use File::Copy;

print "cp META.json MYMETA.json\n";
copy("META.json","MYMETA.json") or die "Copy failed(META.json): $!";

if (-f 'META.yml') {
    print "cp META.yml MYMETA.yml\n";
    copy("META.yml","MYMETA.yml") or die "Copy failed(META.yml): $!";
} else {
    print "There is no META.yml... You may install this module from the repository...\n";
}



src/state.c
#include "state.h"
#include "geohex3.h"
#include "xsutil.h"
#include <string.h>

HV* new_state (pTHX_ const geohex_t *geohex) {
  return init_state(aTHX_ newHV_mortal(), geohex);
}

HV* init_state (pTHX_ HV* state, const geohex_t *geohex) {
  XSUTIL_HV_STORE(state, "lat",   XSUTIL_NEW_SVNV_MORTAL((NV)geohex->location.lat));
  XSUTIL_HV_STORE(state, "lng",   XSUTIL_NEW_SVNV_MORTAL((NV)geohex->location.lng));
  XSUTIL_HV_STORE(state, "x",     XSUTIL_NEW_SVIV_MORTAL((IV)geohex->coordinate.x));
  XSUTIL_HV_STORE(state, "y",     XSUTIL_NEW_SVIV_MORTAL((IV)geohex->coordinate.y));
  XSUTIL_HV_STORE(state, "code",  sv_2mortal(newSVpvn(geohex->code, geohex->level+2)));
  XSUTIL_HV_STORE(state, "level", XSUTIL_NEW_SVUV_MORTAL((UV)geohex->level));
  XSUTIL_HV_STORE(state, "size",  XSUTIL_NEW_SVNV_MORTAL((NV)geohex->size));
  return state;
}

SV* bless_state (pTHX_ const HV* state, const char *class) {
  SV* self = newRV_inc_mortal((SV*)state);
  sv_bless(self, gv_stashpv(class, 1));
  SvREADONLY_on(self);
  return self;
}

geohex_t deflate_to_geohex(pTHX_ HV* state) {
  SV** lat   = XSUTIL_HV_FETCH(state, "lat");
  SV** lng   = XSUTIL_HV_FETCH(state, "lng");
  SV** x     = XSUTIL_HV_FETCH(state, "x");
  SV** y     = XSUTIL_HV_FETCH(state, "y");
  SV** code  = XSUTIL_HV_FETCH(state, "code");
  SV** level = XSUTIL_HV_FETCH(state, "level");
  SV** size  = XSUTIL_HV_FETCH(state, "size");

  geohex_t geohex = {
    .location   = geohex_location((long double)SvNV(*lat), (long double)SvNV(*lng)),
    .coordinate = geohex_coordinate((long)SvIV(*x), (long)SvIV(*y)),
    .level      = (geohex_level_t)SvUV(*level),
    .size       = (long double)SvNV(*size)
  };
  stpcpy(geohex.code, SvPV_nolen(*code));
  return geohex;
}


author/benchmark.pl
use strict;
use warnings;
use utf8;

use Test::More tests => 16;
use Benchmark qw/cmpthese timethese/;
use Geo::Hex::V3;
use Geo::Hex::V3::XS;

my @geohexes = qw/
    XM
    XM4
    XM48
    XM488
    XM4882
    XM48827
    XM488276
    XM4882767
    XM48827674
    XM488276746
    XM4882767460
    XM48827674605
    XM488276746051
    XM4882767460512
    XM48827674605123
    XM488276746051234
/;
my (@latlngs) = map { [Geo::Hex::V3::XS::decode_geohex($_)] } @geohexes;

for my $geohex (@geohexes) {
    my $zone_pp = Geo::Hex::V3::geohex2zone($geohex);
    my $zone_xs = Geo::Hex::V3::XS->new(code => $geohex);
    subtest $geohex => sub {
        is $zone_xs->lat,   $zone_pp->lat,      'lat';
        is $zone_xs->lng,   $zone_pp->lng,      'lng';
        is $zone_xs->x,     $zone_pp->x,        'x';
        is $zone_xs->y,     $zone_pp->y,        'y';
        is $zone_xs->code,  $zone_pp->code,     'code';
        is $zone_xs->level, $zone_pp->level,    'level';
        is $zone_xs->size,  $zone_pp->hex_size, 'size';
    };
}

print '-' x 60, "\n";
print 'Creates zone object from geohex:', "\n";
cmpthese timethese 30000 => {
    pp => sub {
        Geo::Hex::V3::geohex2zone($_) for @geohexes;
    },
    xs => sub {
        Geo::Hex::V3::XS->new(code => $_) for @geohexes;
    },
};
print '-' x 60, "\n";

print '-' x 60, "\n";
print 'Creates zone object from latlng:', "\n";
cmpthese timethese 30000 => {
    pp => sub {
        Geo::Hex::V3::latlng2zone(@$_) for @latlngs;
    },
    xs => sub {
        Geo::Hex::V3::XS->new(lat => $_->[0], lng => $_->[1], level => $_->[2]) for @latlngs;
    },
};
print '-' x 60, "\n";

print '-' x 60, "\n";
print 'Converts latlng to geohex:', "\n";
cmpthese timethese 30000 => {
    pp => sub {
        Geo::Hex::V3::latlng2geohex(@$_) for @latlngs;
    },
    xs => sub {
        Geo::Hex::V3::XS::encode_geohex(@$_) for @latlngs;
    },
};
print '-' x 60, "\n";

print '-' x 60, "\n";
print 'Converts geohex to latlng:', "\n";
cmpthese timethese 30000 => {
    pp => sub {
        Geo::Hex::V3::geohex2latlng($_) for @geohexes;
    },
    xs => sub {
        Geo::Hex::V3::XS::decode_geohex($_) for @geohexes;
    },
};
print '-' x 60, "\n";

__END__
1..16
    # Subtest: XM
    ok 1 - lat
    ok 2 - lng
    ok 3 - x
    ok 4 - y
    ok 5 - code
    ok 6 - level
    ok 7 - size
    1..7
ok 1 - XM
    # Subtest: XM4
    ok 1 - lat
    ok 2 - lng
    ok 3 - x
    ok 4 - y
    ok 5 - code
    ok 6 - level
    ok 7 - size
    1..7
ok 2 - XM4
    # Subtest: XM48
    ok 1 - lat
    ok 2 - lng
    ok 3 - x
    ok 4 - y
    ok 5 - code
    ok 6 - level
    ok 7 - size
    1..7
ok 3 - XM48
    # Subtest: XM488
    ok 1 - lat
    ok 2 - lng
    ok 3 - x
    ok 4 - y
    ok 5 - code
    ok 6 - level
    ok 7 - size
    1..7
ok 4 - XM488
    # Subtest: XM4882
    ok 1 - lat
    ok 2 - lng
    ok 3 - x
    ok 4 - y
    ok 5 - code
    ok 6 - level
    ok 7 - size
    1..7
ok 5 - XM4882
    # Subtest: XM48827
    ok 1 - lat
    ok 2 - lng
    ok 3 - x
    ok 4 - y
    ok 5 - code
    ok 6 - level
    ok 7 - size
    1..7
ok 6 - XM48827
    # Subtest: XM488276
    ok 1 - lat
    ok 2 - lng
    ok 3 - x
    ok 4 - y
    ok 5 - code
    ok 6 - level
    ok 7 - size
    1..7
ok 7 - XM488276
    # Subtest: XM4882767
    ok 1 - lat
    ok 2 - lng
    ok 3 - x
    ok 4 - y
    ok 5 - code
    ok 6 - level
    ok 7 - size
    1..7
ok 8 - XM4882767
    # Subtest: XM48827674
    ok 1 - lat
    ok 2 - lng
    ok 3 - x
    ok 4 - y
    ok 5 - code
    ok 6 - level
    ok 7 - size
    1..7
ok 9 - XM48827674
    # Subtest: XM488276746
    ok 1 - lat
    ok 2 - lng
    ok 3 - x
    ok 4 - y
    ok 5 - code
    ok 6 - level
    ok 7 - size
    1..7
ok 10 - XM488276746
    # Subtest: XM4882767460
    ok 1 - lat
    ok 2 - lng
    ok 3 - x
    ok 4 - y
    ok 5 - code
    ok 6 - level
    ok 7 - size
    1..7
ok 11 - XM4882767460
    # Subtest: XM48827674605
    ok 1 - lat
    ok 2 - lng
    ok 3 - x
    ok 4 - y
    ok 5 - code
    ok 6 - level
    ok 7 - size
    1..7
ok 12 - XM48827674605
    # Subtest: XM488276746051
    ok 1 - lat
    ok 2 - lng
    ok 3 - x
    ok 4 - y
    ok 5 - code
    ok 6 - level
    ok 7 - size
    1..7
ok 13 - XM488276746051
    # Subtest: XM4882767460512
    ok 1 - lat
    ok 2 - lng
    ok 3 - x
    ok 4 - y
    ok 5 - code
    ok 6 - level
    ok 7 - size
    1..7
ok 14 - XM4882767460512
    # Subtest: XM48827674605123
    ok 1 - lat
    ok 2 - lng
    ok 3 - x
    ok 4 - y
    ok 5 - code
    ok 6 - level
    ok 7 - size
    1..7
ok 15 - XM48827674605123
    # Subtest: XM488276746051234
    ok 1 - lat
    ok 2 - lng
    ok 3 - x
    ok 4 - y
    ok 5 - code
    ok 6 - level
    ok 7 - size
    1..7
ok 16 - XM488276746051234
------------------------------------------------------------
Creates zone object from geohex:
Benchmark: timing 30000 iterations of pp, xs...
        pp: 21 wallclock secs (20.84 usr +  0.10 sys = 20.94 CPU) @ 1432.66/s (n=30000)
        xs:  2 wallclock secs ( 1.84 usr +  0.01 sys =  1.85 CPU) @ 16216.22/s (n=30000)
      Rate    pp    xs
pp  1433/s    --  -91%
xs 16216/s 1032%    --
------------------------------------------------------------
------------------------------------------------------------
Creates zone object from latlng:
Benchmark: timing 30000 iterations of pp, xs...
        pp: 22 wallclock secs (21.64 usr +  0.10 sys = 21.74 CPU) @ 1379.94/s (n=30000)
        xs:  2 wallclock secs ( 1.90 usr +  0.00 sys =  1.90 CPU) @ 15789.47/s (n=30000)
      Rate    pp    xs
pp  1380/s    --  -91%
xs 15789/s 1044%    --
------------------------------------------------------------
------------------------------------------------------------
Converts latlng to geohex:
Benchmark: timing 30000 iterations of pp, xs...
        pp: 21 wallclock secs (21.49 usr +  0.10 sys = 21.59 CPU) @ 1389.53/s (n=30000)
        xs:  1 wallclock secs ( 0.59 usr +  0.00 sys =  0.59 CPU) @ 50847.46/s (n=30000)
      Rate    pp    xs
pp  1390/s    --  -97%
xs 50847/s 3559%    --
------------------------------------------------------------
------------------------------------------------------------
Converts geohex to latlng:
Benchmark: timing 30000 iterations of pp, xs...
        pp: 21 wallclock secs (21.22 usr +  0.09 sys = 21.31 CPU) @ 1407.79/s (n=30000)
        xs:  1 wallclock secs ( 0.70 usr +  0.01 sys =  0.71 CPU) @ 42253.52/s (n=30000)
      Rate    pp    xs
pp  1408/s    --  -97%
xs 42254/s 2901%    --
------------------------------------------------------------


builder/MyBuilder.pm
package builder::MyBuilder;
use strict;
use warnings;
use parent qw/Module::Build::XSUtil/;

use Cwd::Guard ();
use File::Spec;
use File::Which;

sub new {
    my ($self, %args) = @_;

    if ( $^O =~ m!(?:MSWin32|cygwin)! ) {
        print "This module does not support Windows.\n";
        exit 0;
    }

    unless (which 'cmake') {
        print "This module require cmake.\n";
        exit 0;
    }

    $args{extra_compiler_flags} = ['-I' . File::Spec->rel2abs(File::Spec->catdir('deps', 'c-geohex3', 'local', 'include'))];
    $args{extra_linker_flags} = [File::Spec->rel2abs(File::Spec->catdir('deps', 'c-geohex3', 'local', 'lib', 'libgeohex3.a'))];

    return $self->SUPER::new(%args);
}

sub ACTION_code {
    my $self = shift;

    {
        my $guard = Cwd::Guard::cwd_guard(File::Spec->catdir('deps', 'c-geohex3'));
        system('cmake', '-DCMAKE_INSTALL_PREFIX=local', '.') == 0 or die;
        system('make') == 0 or die;
        system('make', 'install') == 0 or die;
    }

    return $self->SUPER::ACTION_code(@_);
}

1;


lib/Geo/Hex/V3/XS.pm
package Geo::Hex::V3::XS;
use 5.008001;
use strict;
use warnings;

our $VERSION = "0.10";
use Exporter 5.57 qw/import/;
our @EXPORT_OK = qw/encode_geohex decode_geohex geohex_hexsize/;

use XSLoader;
XSLoader::load(__PACKAGE__, $VERSION);

sub new {
    my ($class, %args) = @_;
    if (exists $args{code}) {
        return $class->_new_with_code($args{code}) || die "invalid code: $args{code}";
    }
    elsif (exists $args{lat} and exists $args{lng} and exists $args{level}) {
        return $class->_new_with_latlng(@args{qw/lat lng level/});
    }
    elsif (exists $args{x} and exists $args{y} and exists $args{level}) {
        return $class->_new_with_xy(@args{qw/x y level/});
    }
    else {
        die "Usage: $class->new(code => \$geohex_code) or $class->new(lat => \$lat, lng => \$lng, lng => \$level) or $class->new(x => \$x, y => \$y, lng => \$level)";
    }
}

1;
__END__

=for stopwords c-geohex3 geohex geohex's

=encoding utf-8

=head1 NAME

Geo::Hex::V3::XS - GeoHex implementation with XS. (c-geohex3 Perl5 binding.)

=head1 SYNOPSIS

    use Geo::Hex::V3::XS;

    my $zone = Geo::Hex::V3::XS->new(code => 'XM488276746');

    # or

    my $zone = Geo::Hex::V3::XS->new(lat => 35.579826, lng => 139.654524, level => 9);

    # or

    my $zone = Geo::Hex::V3::XS->new(x => 11554, y => -3131, level => 7);

    say 'geohex.code:  ', $zone->code;
    say 'geohex.lat:   ', $zone->lat;
    say 'geohex.lng:   ', $zone->lng;
    say 'geohex.level: ', $zone->level;

=head1 DESCRIPTION

Geo::Hex::V3::XS is L<GeoHex|http://geohex.net/> implementation.

=head1 FUNCTIONS

=over

=item C<my $geohex_code = encode_geohex($lat, $lng, $level)>

Convert location,level to geohex's code.

=item C<my ($lat, $lng, $code) = decode_geohex($geohex_code)>

Convert geohex's code to location,level.
This location is center of geohex.

=item C<my $size = geohex_hexsize($level)>

Calculate geohex hex size by level.

=back

=head1 METHODS

=over

=item C<my $zone = Geo::Hex::V3::XS-E<gt>new(...)>

Create geohex zone object.

Arguments can be:

=over

=item * C<code>

Create geohex zone object from geohex code.

    use Geo::Hex::V3::XS;

    my $zone = Geo::Hex::V3::XS->new(code => 'XM488548');

=item * C<lat/lng/level>

Create geohex zone object from location with level.

    use Geo::Hex::V3::XS;

    my $zone = Geo::Hex::V3::XS->new(
        lat   => 40.5814792855475,
        lng   => 134.296601127877,
        level => 7,
    );

=item * C<x/y/level>

Create geohex zone object from coordinate with level.

    use Geo::Hex::V3::XS;

    my $zone = Geo::Hex::V3::XS->new(
        x     => 11554,
        y     => -3131,
        level => 7,
    );

=back

=item C<$zone-E<gt>lat>

Get geohex center location latitude.

=item C<$zone-E<gt>lng>

Get geohex center location longitude.

=item C<$zone-E<gt>x>

Get geohex center x coordinate.

=item C<$zone-E<gt>y>

Get geohex center y coordinate.

=item C<$zone-E<gt>code>

Get geohex code.

=item C<$zone-E<gt>level>

Get geohex level. (0-15)

=item C<$zone-E<gt>size>

Get geohex size.

=item C<my @locations = $zone-E<gt>polygon()>

Get vertex locations of a geohex polygon.

=back

=head1 LICENSE

Copyright (C) 2015 karupanerura <karupa@cpan.org>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

=head1 ALGORITHM LICENSE

Copyright (c) 2009 @sa2da (http://twitter.com/sa2da)
http://www.geohex.org
Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

=head1 AUTHOR

karupanerura E<lt>karupa@cpan.orgE<gt>

=cut

